openmls/src/group/public_group/validation.rs

```diff
@@ -107,129 +110,244 @@ impl PublicGroup {
+    /// - ValSem106
+    /// - ValSem109
+    pub(crate) fn validate_capabilities(
+        &self,
+        proposal_queue: &ProposalQueue,
+    ) -> Result<(), ProposalValidationError> {
+        // ValSem106/ValSem109: Check the required capabilities of the add & update
+        // proposals This includes the following checks:
+        // - Are ciphersuite & version listed in the `Capabilities` Extension?
+        // - If a `RequiredCapabilitiesExtension` is present in the group: Is
+        //   this upported by the node?
+        // - Check that all extensions are contained in the capabilities.
+        // - Check that the capabilities contain the leaf node's credential
+        //   type.
+        // - Check that the credential type is supported by all members of the
+        //   group.
+        // - Check that the capabilities field of this LeafNode indicates
+        //   support for all the credential types currently in use by other
+        //   members.
+
+        // Extract the leaf nodes from the add & update proposals
+        let leaf_nodes = proposal_queue
+            .queued_proposals()
+            .filter_map(|p| match p.proposal() {
+                Proposal::Add(add_proposal) => Some(add_proposal.key_package().leaf_node()),
+                Proposal::Update(update_proposal) => Some(update_proposal.leaf_node()),
+                _ => None,
+            });
+
+        let mut group_leaf_nodes = self.treesync().full_leaves();
+
+        for leaf_node in leaf_nodes {
             // Check if the ciphersuite and the version of the group are
             // supported.
-            let capabilities = add_proposal
-                .add_proposal()
-                .key_package()
-                .leaf_node()
-                .capabilities();
+            let capabilities = leaf_node.capabilities();
             if !capabilities
                 .ciphersuites()
                 .contains(&VerifiableCiphersuite::from(self.ciphersuite()))
                 || !capabilities.versions().contains(&self.version())
             {
-                log::error!("Tried to commit an Add proposal, where either the group's `Ciphersuite` or the group's `ProtocolVersion` is not in the `KeyPackage`'s `Capabilities`.");
                 return Err(ProposalValidationError::InsufficientCapabilities);
             }
+
             // If there is a required capabilities extension, check if that one
             // is supported.
             if let Some(required_capabilities) =
                 self.group_context().extensions().required_capabilities()
             {
                 // Check if all required capabilities are supported.
-                if !capabilities.supports_required_capabilities(required_capabilities) {
-                    log::error!("Tried to commit an Add proposal, where the `Capabilities` of the given `KeyPackage` do not fulfill the `RequiredCapabilities` of the group.");
```

Comment:
We haven't talked about our logging strategy for a while. It might be helpful to log who is missing which capability. Either here or in the `supports_required_capabilities` function. Same for the errors below.

Reply:
I removed outdated logging here, that's all. So far we don't have any logging strategy beyond sporadically logging errors in the context of debugging the library. Happy to discuss more, but maybe in another PR?